### PR TITLE
Make gradle build finished logic CC compatible

### PR DESCRIPTION
**Pull request by breskeby**
The buildfinished build logic in Gradle is not configuration cache. compatible. To allow CI builds running builds with configuration cache enabled we need to move away from using this API to newer APIs provided by Gradle. 

Gradle Configuration Cache: https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:intro 

This generally allows us to run builds with configuration cache enabled on our CI infrastructure as all builds use this build finish logic.

### CODE REVIEW

**Comment by breskeby** [comment]
related to https://github.com/elastic/elasticsearch/issues/57918

File: build-tools-internal/src/main/groovy/elasticsearch.build-complete.gradle [hunk 1]
```diff
@@ -8,86 +8,10 @@
...
-        if (artifactUuid) {
-          buildScan.link 'Artifact Upload', "https://buildkite.com/organizations/elastic/pipelines/${System.getenv('BUILDKITE_PIPELINE_SLUG')}/builds/${buildNumber}/jobs/${System.getenv('BUILDKITE_JOB_ID')}/artifacts/${artifactUuid}"
-        }
-      } catch (Exception e) {
-        logger.lifecycle("Failed to upload buildkite artifact", e)
-      }
-    }
-  }
-}
+project.getPlugins().apply(ElasticsearchBuildFinishedPlugin.class)
```
Comment (mark-vieira) [inline build-tools-internal/src/main/groovy/elasticsearch.build-complete.gradle:17]: Can we just apply this plugin to the root project and ditch this pre-compiled script plugin altogether?

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            : System.getenv("BUILDKITE_BUILD_NUMBER");
+        String performanceTest = System.getenv("BUILD_PERFORMANCE_TEST");
+        if (buildNumber != null && performanceTest == null && GradleUtils.isIncludedBuild(target) == false) {
+            File targetFile = target.file("build/" + buildNumber + ".tar.bz2");
+            File projectDir = target.getProjectDir();
+            File gradleWorkersDir = new File(target.getGradle().getGradleUserHomeDir(), "workers/");
+            BuildScanExtension extension = target.getExtensions().getByType(BuildScanExtension.class);
+            File daemonsLogDir = new File(target.getGradle().getGradleUserHomeDir(), "daemon/" + target.getGradle().getGradleVersion());
+
+            getFlowScope().always(BuildFinishedFlowAction.class, spec -> {
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java:64]: What on earth is a `FlowAction` and how does this equate to a "build finished"? I can't see anythere here that defines _when_ this action runs.

Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java:64]: Don't get me started. That is the last half baked overly complicated gradle api that allows code executing on buildFinished. The when is declared implicitly by this line where we map the BuildWorkResult which implicitly defers this code execution till that result is available. There is currently only `getFlowProviders().getBuildWorkResult()` as "flow provider". It's been around for some versions now, obviously incubating but all that is available right now.

```                
spec.getParameters().getFilteredFiles().addAll(getFlowProviders().getBuildWorkResult().map((result) -> {
```

File: build-tools-internal/src/main/groovy/elasticsearch.build-complete.gradle (hunk 1 above)
Comment (breskeby) [reply build-tools-internal/src/main/groovy/elasticsearch.build-complete.gradle]: fixed

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java (hunk 2 above)
Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java:64]: So calling `getBuildWorkResult()` implicitly makes this a "build finished" hook. That's not confusing...

Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java:64]: it is. but yes.

**Pull request merged**